### ISSUE TITLE
Select Shaders/Colormaps In UI Panel

### ISSUE THREAD

**Issue by arthurfeeney** [google/neuroglancer#243]
I am using Neuroglancer's Python wrapper and I have several of my own colormaps/shaders for an ImageLayer.
I would like to add a drop-down menu (or something similar) to the rendering tab of the layer's UI panel, where the user can see the names of colormaps/shaders and easily pick the one they want.

Some colormaps/shaders may hide or show different information, so having an easy way to pick from existing ones would be great.

I have thought of two ways to change between these colormaps, but neither are quite what I want:
1. create an Action that, on keypress, will cycle through several shaders that each implement a colormap.
2. Give each colormap some ID and use the #uicontrol slider to set the ID corresponding to the desired colormap. All colormaps are defined in the single shader. Something like this:
~~~~
#uicontrol int colormapId slider(min=0, max=5,step=1)
vec4 colormap(float x) {
    if (colormapId == 0) return colormap0(x);
    if (colormapId == 1) return colormap1(x);
    ....
}
....
~~~~

These work, but hide some details from the user, like which shader is currently being used and what the options actually are.
  
I spent a while looking and couldn't find anything, but is there currently a way to pick named shaders from the neuroglancer UI that's similar to what I want? If so, is it possible to set this from the python wrapper?
If not possible, any recommendations on how to add a feature like this would be appreciated!

**Comment by jbms** [comment]
A better UI for selecting colormaps would indeed be nice, and has been requested previously.

If you are interested in implementing this functionality, I would suggest adding a new ShaderColormapControl to neuroglancer/webgl/shader_ui_controls.ts.  For example:

```
#uicontrol colormap myColormap(default = "grayscale")
```

This would be a slight departure from the current syntax:
```
#uicontrol <glsl_type> <name> <control_type>(<parameters>)
```

This would get translated to something like:

```glsl
vec3 myColormap(float value) { ... }
```

Although the current "colormapJet" and "colormapCubehelix" colormaps are defined as GLSL source code, for the purpose of this UI control, it would probably be simpler to instead define them as [256][3] arrays of uint8 values, and bind them as textures.

**Comment by arthurfeeney** [comment]
I just got to look into this and I have some questions:

1. Do you think this should _only_ allow selecting from neuroglancer's built-in grayscale, colormapJet, and colormapHelix? So, the only argument would be `default="colormapName"`.

2. Or, should users should be able to specify their own colormaps implemented in the shader? If so, do you have a recommendation or preference for how they should be passed in to the uicontrol? I.e., it seems like uicontrols currently support numeric and string arguments. This feature would probably require adding a List, but it could be done not-so-elegantly with a delimited string. Probably something like
~~~~
#uicontrol colormap myColormap(default="grayscale", options=["grayscale", "myRedColor", myBlueColor])
~~~~

I'm personally leaning towards only implementing 1 (at least at first). I think neuroglancer should (generally) have good enough colormaps that users should ideally not need to roll their own even if they want to change between them, right?
Thanks!

**Comment by jbms** [comment]
Yes, I agree with you that just implementing 1 makes sense for now.  Once this functionality is in place additional standard colormaps (e.g. from matplotlib) could be added.

One interesting thing to support would be plain colors (maps to `value * color`), or pairs of colors, so grayscale could be equivalent to "black:white".  However, that could be added as an extension later.
